### PR TITLE
Consolidate Container Properties cache in Client Instance

### PR DESCRIPTION
**Pull request by bambriz**
# Description
This PR makes the cache for Container Properties (namely Partition Key Definition and RID) for a container to be cached in the client instance instead of the container instance. 
Because at the moment the Python SDK does not handle caches the same as other SDKs which include refreshing the cache 
(See [Issue 35294](https://github.com/Azure/azure-sdk-for-python/issues/35294)) it is not a regression since it is just consolidating the cache of the properties to a single cache in the client instance and the python sdk has not yet implemented caching the container properties as the other SDKs do. 

This PR will avoid a situation where: if you call single instances of a container multiple times from a client, it will perform a container read for every operation that does Create, Replace, Upsert, and ReadOffers. It will now instead use the cached container properties after the first container read has already been performed for that specific container.

The Client keeps a cache of properties of each container that is used in the lifetime of the client instance. 

The current properties that are cached are "_self" which is the container link "dbsname/coll/containername" , the RID of the container, and the partition key definition of the container.
```python
container_properties = { "_self": "dbsname/colls/containername", "_rid": "########", "partitionKey": {...} }
```

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 1]
```diff
@@ -10,6 +10,7 @@
 #### Bugs Fixed
 * Fixed a bug where change feed query in Async client was not returning all pages due to case-sensitive response headers. See [PR 35090](https://github.com/Azure/azure-sdk-for-python/pull/35090)
 * Fixed a bug when a retryable exception occurs in the first page of a query execution causing query to return 0 results. See [PR 35090](https://github.com/Azure/azure-sdk-for-python/pull/35090).
+* Consolidated Container Properties Cache to be in the Client to cache partition key definition and container rid to avoid uncessary container reads. See [PR35293](https://github.com/Azure/azure-sdk-for-python/pull/35293/files)
```
Comment (philipthomas-MSFT) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:13]: unnecessary

Comment (bambriz) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md:13]: This Pr does specifically tackle on an issue where we were calling container reads more than we needed to. I could rephrase it to just mention it does fix that issue.

Comment (philipthomas-MSFT) [reply sdk/cosmos/azure-cosmos/CHANGELOG.md:13]: LOL. Apologies, I meant the spelling of **unnecessary** was wrong.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py [hunk 2]
```diff
@@ -3272,11 +3271,12 @@ def _UpdateSessionIfRequired(
...
-        if collection_link in self.partition_key_definition_cache:
-            partition_key_definition = self.partition_key_definition_cache.get(collection_link)
+        if collection_link in self.collection_properties_cache:
+            cached_collection: Dict[str, Any] = self.collection_properties_cache.get(collection_link)
+            partition_key_definition = cached_collection.get("partitionKey")
         # Else read the collection from backend and add it to the cache
         else:
             collection = self.ReadContainer(collection_link)
             partition_key_definition = collection.get("partitionKey")
-            self.partition_key_definition_cache[collection_link] = partition_key_definition
```
Comment (philipthomas-MSFT) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py:3281]: Can you include somewhere on the description what the key, **collection_link**, and value, **collection**, expects to look like so I can assess the size?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 3]
```diff
@@ -168,11 +169,11 @@ async def read(
         if populate_quota_info is not None:
             request_options["populateQuotaInfo"] = populate_quota_info
 
-        collection_link = self.container_link
-        self._properties = await self.client_connection.ReadContainer(
-            collection_link, options=request_options, **kwargs
+        coll_link = self.container_link
+        self.client_connection.collection_properties_cache[coll_link] = await self.client_connection.ReadContainer(
+            coll_link, options=request_options, **kwargs
```
Comment (philipthomas-MSFT) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:174]: When I see things like **coll_link** and **container_link**, I immediately think that these are two different things. When in fact, **collection** and **container** are the same.

File: sdk/cosmos/azure-cosmos/test/test_container_properties_cache.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+                """
+        try:
+            func(*args, **kwargs)
+            self.assertFalse(True, 'function should fail.')
+        except exceptions.CosmosHttpResponseError as inst:
+            self.assertEqual(inst.status_code, status_code)
+
+    @classmethod
+    def setUpClass(cls):
+        if (cls.masterKey == '[YOUR_KEY_HERE]' or
```
Comment (philipthomas-MSFT) [inline sdk/cosmos/azure-cosmos/test/test_container_properties_cache.py:46]: for these tests, are you storing **masterKey** and **host** as an environment variable? Is this test ran gated?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 5]
```diff
@@ -168,11 +169,11 @@ async def read(
         if populate_quota_info is not None:
             request_options["populateQuotaInfo"] = populate_quota_info
 
-        collection_link = self.container_link
-        self._properties = await self.client_connection.ReadContainer(
-            collection_link, options=request_options, **kwargs
+        coll_link = self.container_link
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:172]: This line isn't necessary right? We can just use `self.container_link` directly in the lines below?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 6]
```diff
@@ -81,19 +81,20 @@ def __init__(
...
 
     def __repr__(self) -> str:
         return "<ContainerProxy [{}]>".format(self.container_link)[:1024]
 
     async def _get_properties(self) -> Dict[str, Any]:
-        if self._properties is None:
-            self._properties = await self.read()
-        return self._properties
+        if self.container_link not in self.client_connection.collection_properties_cache:
+            self.client_connection.collection_properties_cache[self.container_link] = await self.read()
```
Comment (annatisch) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:96]:
```suggestion
            await self.read()
```
the `read()` method is already updating the cache, so we don't need to do it here as well.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py (hunk 5 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:172]: Yes, but I did this to avoid having to put a pylint ignore comment for having a line of code that is too long.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py (hunk 3 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py]: I updated to keep it as consistently as possible as container within the scope of my changes

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py [hunk 7]
```diff
@@ -83,17 +83,21 @@ def __init__(
         self.id = id
         self.container_link = "{}/colls/{}".format(database_link, self.id)
         self.client_connection = client_connection
-        self._properties = properties
         self._is_system_key: Optional[bool] = None
         self._scripts: Optional[ScriptsProxy] = None
+        if properties:
+            self.client_connection.container_properties_cache[self.container_link] = {
+                "_self": properties.get("_self", None), "_rid": properties.get("_rid", None),
```
Comment (ealsur) [inline sdk/cosmos/azure-cosmos/azure/cosmos/container.py:90]: nit: The duplication in the logic here and in line 182 and in the other file on which properties to extract can lead to in the future, to someone updating 1 place only instead of all 4. Would it be better to have a shared function that creates the cached entry based on the input and share that function across? So in the future, if you want to add more properties to the cache, you touch 1 place and not 4.

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py [hunk 8]
```diff
@@ -173,11 +177,13 @@ def read(  # pylint:disable=docstring-missing-param
...
         if populate_quota_info is not None:
             request_options["populateQuotaInfo"] = populate_quota_info
-        collection_link = self.container_link
-        self._properties = self.client_connection.ReadContainer(
-            collection_link, options=request_options, **kwargs
-        )
-        return self._properties
+        container = self.client_connection.ReadContainer(self.container_link, options=request_options, **kwargs)
+        # Only cache Container Properties that will not change in the lifetime of the container
+        self.client_connection.container_properties_cache[self.container_link] = {
```
Comment (ealsur) [inline sdk/cosmos/azure-cosmos/azure/cosmos/container.py:182]: `self.container_link` contains the name? If so, how is this PR addressing the recreation scenario? Or is that being done in a separate PR?

Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/container.py:182]: That is being done in a separate pr I am working on. That one should be out soon after this PR is merged.

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py (hunk 7 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/container.py]: Very good point, I made the change to make a single method to populate the properties cache.

**Review by ealsur — APPROVED** [review]
LGTM. Please get Python folks to look at it :)

File: sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py [hunk 9]
```diff
@@ -3270,13 +3269,14 @@ def _UpdateSessionIfRequired(
...
-            partition_key_definition = self.partition_key_definition_cache.get(collection_link)
+        if collection_link in self.container_properties_cache:
+            cached_collection: Dict[str, Any] = self.container_properties_cache.get(collection_link, {})
+            partition_key_definition = cached_collection.get("partitionKey")
         # Else read the collection from backend and add it to the cache
         else:
             collection = self.ReadContainer(collection_link)
             partition_key_definition = collection.get("partitionKey")
-            self.partition_key_definition_cache[collection_link] = partition_key_definition
+            self.container_properties_cache[collection_link] = collection
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py:3281]: should this be _set_properties_cache(collection)?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 10]
```diff
@@ -81,19 +82,20 @@ def __init__(
...
     def __repr__(self) -> str:
         return "<ContainerProxy [{}]>".format(self.container_link)[:1024]
 
     async def _get_properties(self) -> Dict[str, Any]:
-        if self._properties is None:
-            self._properties = await self.read()
-        return self._properties
+        if self.container_link not in self.client_connection.container_properties_cache:
+            await self.read()
+        return self.client_connection.container_properties_cache[self.container_link]
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:98]: for _get_properties -> this should always call self.read()? because we should always return all collection properties? (like index policy)

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py [hunk 11]
```diff
@@ -81,19 +82,20 @@ def __init__(
...
 
     def __repr__(self) -> str:
         return "<ContainerProxy [{}]>".format(self.container_link)[:1024]
 
     async def _get_properties(self) -> Dict[str, Any]:
-        if self._properties is None:
-            self._properties = await self.read()
-        return self._properties
+        if self.container_link not in self.client_connection.container_properties_cache:
+            await self.read()
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:97]: once read, I guess we can also update the cache?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py [hunk 12]
```diff
@@ -3034,13 +3034,14 @@ async def _AddPartitionKey(self, collection_link, document, options):
...
-            partitionKeyDefinition = self.partition_key_definition_cache.get(collection_link)
+        if collection_link in self.container_properties_cache:
+            cached_collection = self.container_properties_cache.get(collection_link)
+            partitionKeyDefinition = cached_collection.get("partitionKey")
         # Else read the collection from backend and add it to the cache
         else:
             collection = await self.ReadContainer(collection_link)
             partitionKeyDefinition = collection.get("partitionKey")
-            self.partition_key_definition_cache[collection_link] = partitionKeyDefinition
+            self.container_properties_cache[collection_link] = collection
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py:3044]: same question as above, should this be _set_properties_cache(collection)? (since we only persist few properties in the cache)

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py [hunk 13]
```diff
@@ -83,17 +84,18 @@ def __init__(
...
     def __repr__(self) -> str:
         return "<ContainerProxy [{}]>".format(self.container_link)[:1024]
 
     def _get_properties(self) -> Dict[str, Any]:
-        if self._properties is None:
-            self._properties = self.read()
-        return self._properties
+        if self.container_link not in self.client_connection.container_properties_cache:
+            self.read()
+        return self.client_connection.container_properties_cache[self.container_link]
```
Comment (xinlian12) [inline sdk/cosmos/azure-cosmos/azure/cosmos/container.py:98]: same question: _get_properties should always translate into self.read?

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py (hunk 10 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:98]: For properties that might change like index policy it might be better to get the properties from a read, with this cache we reuse the ones we wont have to change to avoid having to do read calls on them, in particular partition key definition would be one the ones called the most.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py (hunk 11 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:97]: Yes, the cache is updated in the read itself. Whenever a read is called the cache will be updated.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py (hunk 12 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_cosmos_client_connection_async.py:3044]: Yes, thank you for catching that

File: sdk/cosmos/azure-cosmos/azure/cosmos/container.py (hunk 13 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/container.py:98]: That is only when we first populate the cache, in the container itself the properties we are most actively using are rid, partition key definition, and container link, since those don't change, it saves resources by caching them on the first container read which might happen before _get_properties is called, in which case we don't need to do another read. Even if we do cache all the properties, we would just end up needing to do a read anyways for the other ones.

File: sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py (hunk 9 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/_cosmos_client_connection.py:3281]: yes, thank you for catching that.

File: sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py (hunk 10 above)
Comment (xinlian12) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:98]: what are the returns for _get_properties as of today? does it include all properties or just partitionKeyDefinition?

Comment (bambriz) [reply sdk/cosmos/azure-cosmos/azure/cosmos/aio/_container.py:98]: As of today it essentially returns the container dictionary that was cached from the container instance. This includes everything you would have gotten from calling a read for the container, but won't have the latest information if some properties had changed, and only lasts for the lifetime of the container instance.

File: sdk/cosmos/azure-cosmos/CHANGELOG.md [hunk 14]
```diff
@@ -10,6 +10,7 @@
 #### Bugs Fixed
 * Fixed a bug where change feed query in Async client was not returning all pages due to case-sensitive response headers. See [PR 35090](https://github.com/Azure/azure-sdk-for-python/pull/35090)
 * Fixed a bug when a retryable exception occurs in the first page of a query execution causing query to return 0 results. See [PR 35090](https://github.com/Azure/azure-sdk-for-python/pull/35090).
+* Consolidated Container Properties Cache to be in the Client to cache partition key definition and container rid to avoid unnecessary container reads. See [PR35293](https://github.com/Azure/azure-sdk-for-python/pull/35293/files)
```
Comment (kushagraThapar) [inline sdk/cosmos/azure-cosmos/CHANGELOG.md:13]:
```suggestion
* Consolidated Container Properties Cache to be in the Client to cache partition key definition and container rid to avoid unnecessary container reads. See [PR 35293](https://github.com/Azure/azure-sdk-for-python/pull/35293)
```

**Review by kushagraThapar — CHANGES_REQUESTED** [review]
Thanks for the changes @bambriz. Overall, they look good to me. 
I have 2 concerns: 
1. We should add tests which also make sure the cache is populated correctly, i.e. with correct properties like rid and partition key definition, etc.
2. Currently it looks like the `container_properties_cache` is public, which should not be the case. AFAIK, python doesn't have private variables, but at least we can use naming conventions to make it private?

**Comment by bambriz** [comment]
> Thanks for the changes @bambriz. Overall, they look good to me. I have 2 concerns:
> 
> 1. We should add tests which also make sure the cache is populated correctly, i.e. with correct properties like rid and partition key definition, etc.
> 2. Currently it looks like the `container_properties_cache` is public, which should not be the case. AFAIK, python doesn't have private variables, but at least we can use naming conventions to make it private?

Access to the properties cache will be needed outside of the cosmos client connection class, so we would need public setters and (Read only) getters for the cache. I'll go ahead and make those changes.

File: sdk/cosmos/azure-cosmos/test/test_container_properties_cache_async.py [hunk 15]
```diff
@@ -0,0 +1,84 @@
...
+
+    async def test_container_properties_cache(self):
+        client = self.client
+        database_name = "Container Properties Cache Test DB Async " + str(uuid.uuid4())
+        created_db = await client.create_database(database_name)
+        container_name = str(uuid.uuid4())
+        container_pk = "PK"
+        # Create The Container
+        try:
+            await client.get_database_client(database_name).create_container(id=container_name, partition_key=PartitionKey(
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_container_properties_cache_async.py:55]: any reason why you are using the database proxy here instead of the created_db from the create operation?

File: sdk/cosmos/azure-cosmos/test/test_container_properties_cache.py [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+        """
+
+    configs = test_config.TestConfig
+    host = configs.host
+    masterKey = configs.masterKey
+    connectionPolicy = configs.connectionPolicy
+    last_headers = []
+    client: cosmos_client.CosmosClient = None
+
+    def __AssertHTTPFailureWithStatus(self, status_code, func, *args, **kwargs):
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_container_properties_cache.py:31]: is this ever used in the test?

File: sdk/cosmos/azure-cosmos/test/test_retry_policy.py [hunk 17]
```diff
@@ -241,12 +241,6 @@ def test_default_retry_policy_for_create(self):
 
             self.assertDictEqual(created_document, {})
 
-            # 3 retries for readCollection. No retry for createDocument.
-            # Counter ends up in three additional calls while doing create_item,
-            # which are reads to the database and container before creating the item.
```
Comment (simorenoh) [inline sdk/cosmos/azure-cosmos/test/test_retry_policy.py:246]: very random but what happened with this test?

**Review by simorenoh — COMMENTED** [review]
just a couple questions, looks good overall thanks Bryan!

**Review by kushagraThapar — APPROVED** [review]
LGTM thanks @bambriz

File: sdk/cosmos/azure-cosmos/test/test_retry_policy.py (hunk 17 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/test/test_retry_policy.py:246]: This PR actually fixes this issue and caused it to fail, the reason there were 3 additional calls were because of the old cache that was at the container level. There were a couple fo other tests that took into account the additional read calls as well.

File: sdk/cosmos/azure-cosmos/test/test_container_properties_cache_async.py (hunk 15 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/test/test_container_properties_cache_async.py:55]: Its to check the cache while using the proxies like this instead of an instance like how it is usually done, since that's the scenario in which the old cache had issues being used. I will also be expanding this tests in the followup PR.

File: sdk/cosmos/azure-cosmos/test/test_container_properties_cache.py (hunk 16 above)
Comment (bambriz) [reply sdk/cosmos/azure-cosmos/test/test_container_properties_cache.py:31]: I do not believe so, thanks for catching this.

**Review by simorenoh — APPROVED** [review]
Thanks for replying, makes sense! Thanks!
